Recovering from Loss on Development_2 Branch

SwiffterAnim opened the pull request:
Starting by the fix of movement when player gets hit.

SwiffterAnim committed:
Recovering from Loss on Development_2 Branch

Starting by the fix of movement when player gets hit.

SwiffterAnim committed:
Protective Shield

Added a protective shield when player gets hit to temporarily protect from projectiles.

SwiffterAnim committed:
Enemy Shooter destroyer

Made a way to kill the Shooter, and other bits and bobs.
Really not sure how I organised this.

SwiffterAnim committed:
Implementing Events

Tried implementing more events, for when the shooter gets killed and for when the player gets killed.
There were spots where I was checking on Update() if they were alive or not to do something, and calling that every frame for stuff I only needed once.

omidizadi commented on Assets/Modules/Enemies/ProjectileEntity.cs:
```diff
@@ -0,0 +1,23 @@
+    public Rigidbody2D rb;
+
+    private float timer = 0;
+    public Vector2 thisObjectsVelocity;
```
One tip for naming variables, the name should not have information about the context, because we already have that information. For example here, `velocity` would be enough. because whoever tries to access this value will already know that this value exists in this `ProjectileEntity` instance

SwiffterAnim replied:
Done

omidizadi commented on Assets/Modules/Enemies/EnemyKillController.cs:
```diff
@@ -0,0 +1,12 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class EnemyKillController : MonoBehaviour
```
I would call this class the same as `PlayerLifeController` not killController. The reason is if two classes are doing the same job for different entities (even though the logic is different), we should keep their name as consistent as possible.

SwiffterAnim replied:
Done

omidizadi commented on Assets/Modules/Enemies/EnemyMovementController.cs:
```diff
@@ -0,0 +1,101 @@
+
+public class EnemyMovementController : MonoBehaviour
+{
+    public static event Action OnShooterIsDeadEvent;
```
This event should be in the `EnemyLifeController` because it's not related to movement, but to life.

omidizadi replied:
The logic related to this event should also move to the other script.

SwiffterAnim replied:
Done

omidizadi commented on Assets/Modules/Enemies/ProjectileDestroyController.cs:
```diff
@@ -0,0 +1,34 @@
+{
+
+    private float timer;
+
```
Also, ensure that your VS Code has a C# prettifier plugin and learn how to make the scripts pretty using the intended shortcut. The prettifiers will always remove unnecessary spaces and indents.

SwiffterAnim replied:
ok I found one called CSharpier hope it's good even though I don't  love that they put SerializedField on top and then the variable on the bottom. Otherwise it helped! thanks!

omidizadi commented on Assets/Modules/Enemies/ProjectileSpawnController.cs:
```diff
@@ -15,13 +15,17 @@ public class ProjectileSpawnController : MonoBehaviour
-        timer += Time.deltaTime;
-
-        if (timer > timeToShoot)
+        if(PlayerManager.Instance.isAlive) //Not sure if I should use the OnPlayerDiedEvent or use the Player Manager like this. I guess I always need a book like "canShoot" anyway, even if I use the event?
```
You can find out which one to use by following this trick:
* When do you need to fire your logic? When the enemy dies? The answer here is no therefore the `OnPlayerDiedEvent` is not suitable. This logic needs to be fired/checked every frame when the player is alive. So your implementation works 👍

SwiffterAnim replied:
Thanks hehe

omidizadi commented on Assets/Modules/Enemies/ShooterVisualController.cs:
```diff
@@ -0,0 +1,70 @@
+
+    private void OnEnable()
+    {
+        EnemyMovementController.OnShooterIsDeadEvent += OnShooterIsDeadEvent;
```
Good usage of the dead event 👍 😄 
The event will move to the life controller tho (as I said earlier).

omidizadi commented on Assets/Modules/Enemies/EnemyEntity.cs:
```diff
@@ -0,0 +1,12 @@
+
+public class EnemyEntity : MonoBehaviour
+{
+    public float shooterDestroyerSpeed = 1;
```
Also forgot to mention I see that you've completely understood the concept of entities and you just put raw data in them (no logic) Good job 👍

SwiffterAnim replied:
Haha I've been tempted a few times xD

omidizadi commented on Assets/Modules/Enemies/ShooterVisualController.cs:
```diff
@@ -0,0 +1,70 @@
+            {
+                Instantiate(explosion, destroyer.gameObject.transform.position, Quaternion.identity);
+                CameraManager.Instance.Shake();
+                enemyKillController.DestroyThisEnemy();
```
I'm not sure if we need this extra `DestroyThisEnemy` method. You can just use `Destroy( enemyKillController.gameObject);`

SwiffterAnim replied:
Done

omidizadi commented on Assets/Modules/Health/PlayerLifeController.cs:
```diff
@@ -44,15 +46,15 @@ private void OnDestroy()
     {
         crackIndex = 0;
-        isAlive = true;
+        PlayerManager.Instance.isAlive = true;
```
Oh this is not good. We shouldn't be able to set the value of ` PlayerManager.Instance.isAlive` we should only read from it. So I suggest doing the following thing. Do not use PlayerManager in player scripts in general and make the isAlive variable a property like this in PlayerManager.cs:

```
[Serializefield] private  PlayerLifeController playerLifeController;

public bool IsAlive => playerLifeController.isAlive;
```
This way you are protecting this super accessible variable from getting wrongly set by other classes.

SwiffterAnim replied:
Wait I misunderstood, so you're telling me to create the original isAlive variable on the PlayerLifeController, and access that on the PlayerManager through that line.
I guess I have to put that on Update() for it to be constantly checking? Or I should subscribe to the OnPlayerDiedEvent on the PlayerManager, to change the value of this variable?
Also, you said **"Do not use PlayerManager in player scripts in general"**, does that mean like never check this either?
"if (PlayerManager.Instance.isAlive)"?
Then why do I need this public variable here at all?

SwiffterAnim replied:
Or on any player scripts I check if he's alive through the PlayerLifeController.cs and through other scripts (like enemies) I check through PlayerManager and that's why we change both?

omidizadi reviewed (commented):
Also a comment about the folder structure. I saw that you have a folder called prefabs:
![image](https://github.com/user-attachments/assets/1c88aad7-bbe3-4a6d-8e5f-96e4f6c5992d)
with prefabs from every module in it. It shouldn't exist. You should move all prefabs from this folder to their related module folder

SwiffterAnim committed:
Life Liquid (not sim) and your notes.

Did partially yesterday some of the life logic and fixed all your notes from the PR.

SwiffterAnim committed:
Adding Hooks Visuals and Animation
